### ISSUE TITLE
setCookie issue

### ISSUE THREAD

**Issue by Elaoed** [miyakogi/pyppeteer#94]
Hi: 
        I have a question about use cookie in pyppeteer.  How should we pass cookie when access the website. In requests, there is requests.get(url, cookies=cookies).  I didn't find a optional parameter in page.goto function. But i found it as a function of page class. So i tried like this:

    cookies = {'name': "", 'value': ""}  # just a example
    browser = await launch({"executablePath": "/usr/bin/google-chrome-stable"}, args=['--no-sandbox'], headless=True)
    page = await browser.newPage()
    await page.setCookie(cookies)
    print('after set cookie')
    response = await page.goto('http://example.com', timeout=100000)
    await browser.close()

But below exception was raised:

>   asyncio.get_event_loop().run_until_complete(main())
>   File "/usr/local/python3.6/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
>     return future.result()
>   File "test.py", line 26, in main
>     await page.setCookie(cookies)
>   File "/usr/local/python3.6/lib/python3.6/site-packages/pyppeteer/page.py", line 392, in setCookie
>     await self.deleteCookie(*items)
>   File "/usr/local/python3.6/lib/python3.6/site-packages/pyppeteer/page.py", line 373, in deleteCookie
>     await self._client.send('Network.deleteCookies', item)
>   File "/usr/local/python3.6/lib/python3.6/site-packages/pyppeteer/connection.py", line 200, in send
>     return await callback
> pyppeteer.errors.NetworkError: Protocol Error: Invalid parameters name: string value expected

Could you please help with this and explain a little about  cookie machnism difference between pyppeteer and other http library

**Comment by miyakogi** [comment]
`page.setCookie()` method is not available on blank page.
You need to specify a URL to set cookie in this case.

For example, `cookies = {'url': 'http://example.com', 'name': '', 'value': ''}` will work as you expected.

**Comment by Elaoed** [comment]
Yes. using cookies.update({'url': 'http://example.com', 'name': '', 'value': ''}) makes project works without error, But it seems like cookies doesn't work. because the command screenshot capture the login page once again. And i tried use requests with same cookie and it works.

**Comment by Elaoed** [comment]
Yes. It works and the mechanism is little different with cookie in requests i think. 

[{'name': 'username', 'value': '"testuser"', 'domain': 'example.com', 'path': '/', 'expires': 1531274783.734359, 'size': 121, 'httpOnly': False, 'secure': False, 'session': False}] 
this is the cookie i captured from await page.cookies(). 

{'uesrname': '"testuser"'}
this is the cookie i captured from requests response.cookies

It seems like cookies in pyppeteer has more details

**Comment by miyakogi** [comment]
Pyppeteer's cookie is fully depending on chromium.
I'm not sure why does not work on your site.

**Comment by Elaoed** [comment]
Ok thank you. Actually I use chrome all the time

**Comment by qbss1221** [comment]
I have encountered the same problem, how can you solve it?

**Comment by marcoschi** [comment]
I solved by using the setExtraHTTPHeaders method (https://miyakogi.github.io/pyppeteer/reference.html#pyppeteer.page.Page.setExtraHTTPHeaders):

```python
import asyncio
from pyppeteer import launch

async def download():
    browser = await launch({"headless": True})
    page = await browser.newPage()
    await page.setExtraHTTPHeaders({'Cookie': 'SetCurrency=EUR; lang=it_IT'})
    await page.goto('http://localhost/test.py')
    await page.setExtraHTTPHeaders({}) # use this if you want to reset extra headers
    html = await page.evaluate('() => document.body.innerHTML')
    await page.close()
    await browser.close()
    return html

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(download())
    print(result)
```

**Comment by cexoddd2015** [comment]
setExtraHTTPHeader Only for the next page has effect, after the other pages are not effective, solve

**Comment by cexoddd2015** [comment]
> I solved by using the setExtraHTTPHeaders method (https://miyakogi.github.io/pyppeteer/reference.html#pyppeteer.page.Page.setExtraHTTPHeaders):
> 
> ```python
> import asyncio
> from pyppeteer import launch
> 
> async def download():
>     browser = await launch({"headless": True})
>     page = await browser.newPage()
>     await page.setExtraHTTPHeaders({'Cookie': 'SetCurrency=EUR; lang=it_IT'})
>     await page.goto('http://localhost/test.py')
>     await page.setExtraHTTPHeaders({}) # use this if you want to reset extra headers
>     html = await page.evaluate('() => document.body.innerHTML')
>     await page.close()
>     await browser.close()
>     return html
> 
> if __name__ == "__main__":
>     loop = asyncio.get_event_loop()
>     result = loop.run_until_complete(download())
>     print(result)
> ```

**setExtraHTTPHeaders** Only for the next page has effect, after the other pages are not effective, solve

**Comment by GuoZhaoHui628** [comment]
谢谢，如果没解决可以参考我博客 [https://guozh.net/puppeteer-cookie-login/](url)
